// blockstack/lib/profiles/profileSchemas/organization.js
"use strict";

exports.__esModule = true;
exports.Organization = void 0;

var _schemaInspector = _interopRequireDefault(require("schema-inspector"));

var _profileTokens = require("../profileTokens");

var _profile = require("../profile");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schemaDefinition = {
  type: 'object',
  properties: {
    '@context': {
      type: 'string',
      optional: true
    },
    '@type': {
      type: 'string'
    },
    '@id': {
      type: 'string',
      optional: true
    }
  }
};

class Organization extends _profile.Profile {
  constructor(profile = {}) {
    super(profile);
    this._profile = Object.assign({}, {
      '@type': 'Organization'
    }, this._profile);
  }

  static validateSchema(profile, strict = false) {
    schemaDefinition.strict = strict;
    return _schemaInspector.default.validate(schemaDefinition, profile);
  }

  static fromToken(token, publicKeyOrAddress = null) {
    const profile = (0, _profileTokens.extractProfile)(token, publicKeyOrAddress);
    return new Organization(profile);
  }

}

exports.Organization = Organization;
