// radiks/lib/models/user.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _keys = require("blockstack/lib/keys");

var _encryption = require("blockstack/lib/encryption");

var _model = _interopRequireDefault(require("../model"));

var _signingKey = _interopRequireDefault(require("./signing-key"));

var _groupMembership = _interopRequireDefault(require("./group-membership"));

var _helpers = require("../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const decrypted = true;

let BlockstackUser =
/*#__PURE__*/
function (_Model) {
  _inherits(BlockstackUser, _Model);

  function BlockstackUser() {
    _classCallCheck(this, BlockstackUser);

    return _possibleConstructorReturn(this, _getPrototypeOf(BlockstackUser).apply(this, arguments));
  }

  _createClass(BlockstackUser, [{
    key: "createSigningKey",
    value: async function createSigningKey() {
      const key = await _signingKey.default.create();
      this.attrs.personalSigningKeyId = key._id;
      return key;
    }
  }, {
    key: "sign",
    value: async function sign() {
      this.attrs.signingKeyId = 'personal';
      const {
        appPrivateKey
      } = (0, _helpers.loadUserData)();
      const contentToSign = [this._id];

      if (this.attrs.updatedAt) {
        contentToSign.push(this.attrs.updatedAt);
      }

      const {
        signature
      } = (0, _encryption.signECDSA)(appPrivateKey, contentToSign.join('-'));
      this.attrs.radiksSignature = signature;
      return this;
    }
  }], [{
    key: "currentUser",
    value: function currentUser() {
      if (typeof window === 'undefined') {
        return null;
      }

      const userData = (0, _helpers.loadUserData)();

      if (!userData) {
        return null;
      }

      const {
        username,
        profile,
        appPrivateKey
      } = userData;
      const publicKey = (0, _keys.getPublicKeyFromPrivate)(appPrivateKey);
      const Clazz = this;
      const user = new Clazz({
        _id: username,
        username,
        publicKey,
        profile
      });
      return user;
    }
  }, {
    key: "createWithCurrentUser",
    value: function createWithCurrentUser() {
      return new Promise((resolve, reject) => {
        const resolveUser = (user, _resolve) => {
          user.save().then(() => {
            _groupMembership.default.cacheKeys().then(() => {
              _resolve(user);
            });
          });
        };

        try {
          const user = this.currentUser();
          user.fetch().catch(() => {// console.error('caught error', e);
          }).finally(() => {
            // console.log(user.attrs);
            const userData = (0, _helpers.loadUserData)();
            const {
              username,
              profile,
              appPrivateKey
            } = userData;
            const publicKey = (0, _keys.getPublicKeyFromPrivate)(appPrivateKey);
            user.update({
              username,
              profile,
              publicKey
            });

            if (!user.attrs.personalSigningKeyId) {
              user.createSigningKey().then(key => {
                (0, _helpers.addPersonalSigningKey)(key);
                resolveUser(user, resolve);
              });
            } else {
              _signingKey.default.findById(user.attrs.personalSigningKeyId).then(key => {
                (0, _helpers.addPersonalSigningKey)(key);
                resolveUser(user, resolve);
              });
            }
          });
        } catch (error) {
          reject(error);
        }
      });
    }
  }]);

  return BlockstackUser;
}(_model.default);

exports.default = BlockstackUser;

_defineProperty(BlockstackUser, "className", 'BlockstackUser');

_defineProperty(BlockstackUser, "schema", {
  username: {
    type: String,
    decrypted
  },
  publicKey: {
    type: String,
    decrypted
  },
  profile: {
    type: String,
    decrypted
  },
  personalSigningKeyId: String
});
