// blockstack/lib/auth/authConstants.js
"use strict";

exports.__esModule = true;
exports.LOCALSTORAGE_SESSION_KEY = exports.NAME_LOOKUP_PATH = exports.DEFAULT_CORE_NODE = exports.BLOCKSTACK_DEFAULT_GAIA_HUB_URL = exports.BLOCKSTACK_APP_PRIVATE_KEY_LABEL = exports.DEFAULT_SCOPE = exports.DEFAULT_BLOCKSTACK_HOST = exports.BLOCKSTACK_STORAGE_LABEL = exports.BLOCKSTACK_HANDLER = void 0;
const BLOCKSTACK_HANDLER = 'blockstack';
exports.BLOCKSTACK_HANDLER = BLOCKSTACK_HANDLER;
const BLOCKSTACK_STORAGE_LABEL = 'blockstack';
exports.BLOCKSTACK_STORAGE_LABEL = BLOCKSTACK_STORAGE_LABEL;
const DEFAULT_BLOCKSTACK_HOST = 'https://browser.blockstack.org/auth';
exports.DEFAULT_BLOCKSTACK_HOST = DEFAULT_BLOCKSTACK_HOST;
const DEFAULT_SCOPE = ['store_write'];
exports.DEFAULT_SCOPE = DEFAULT_SCOPE;
const BLOCKSTACK_APP_PRIVATE_KEY_LABEL = 'blockstack-transit-private-key';
exports.BLOCKSTACK_APP_PRIVATE_KEY_LABEL = BLOCKSTACK_APP_PRIVATE_KEY_LABEL;
const BLOCKSTACK_DEFAULT_GAIA_HUB_URL = 'https://hub.blockstack.org';
exports.BLOCKSTACK_DEFAULT_GAIA_HUB_URL = BLOCKSTACK_DEFAULT_GAIA_HUB_URL;
const DEFAULT_CORE_NODE = 'https://core.blockstack.org';
exports.DEFAULT_CORE_NODE = DEFAULT_CORE_NODE;
const NAME_LOOKUP_PATH = '/v1/names';
exports.NAME_LOOKUP_PATH = NAME_LOOKUP_PATH;
const LOCALSTORAGE_SESSION_KEY = 'blockstack-session';
exports.LOCALSTORAGE_SESSION_KEY = LOCALSTORAGE_SESSION_KEY;
