// blockstack/lib/auth/sessionData.js
"use strict";

exports.__esModule = true;
exports.SessionData = void 0;

var _errors = require("../errors");

const SESSION_VERSION = '1.0.0';

class SessionData {
  // required after sign in
  // required after sign in
  // required after sign in
  // using this in place of
  // window.localStorage.setItem(BLOCKSTACK_STORAGE_LABEL, JSON.stringify(userData))
  constructor(options) {
    this.version = SESSION_VERSION;
    this.appPrivateKey = options.appPrivateKey;
    this.identityAddress = options.identityAddress;
    this.username = options.username;
    this.coreNode = options.coreNode;
    this.hubUrl = options.hubUrl;
    this.userData = options.userData;
    this.transitKey = options.transitKey; // initializing Gaia connection requires a network request
    // so we'll defer it until the first time it's needed

    this.gaiaHubConfig = null;
  }

  getGaiaHubConfig() {
    return this.gaiaHubConfig;
  }

  setGaiaHubConfig(config) {
    this.gaiaHubConfig = config;
  }

  static fromJSON(json) {
    if (json.version !== SESSION_VERSION) {
      throw new _errors.InvalidStateError(`JSON data version ${json.version} not supported by SessionData`);
    }

    const options = {
      appPrivateKey: json.appPrivateKey,
      identityAddress: json.identityAddress,
      username: json.username,
      coreNode: json.coreNode,
      hubUrl: json.hubUrl,
      userData: json.userData,
      transitKey: json.transitKey
    };
    return new SessionData(options);
  }

  toString() {
    return JSON.stringify(this);
  }

}

exports.SessionData = SessionData;
