// blockstack/lib/profiles/profile.js
"use strict";

exports.__esModule = true;
exports.Profile = void 0;

var _schemaInspector = _interopRequireDefault(require("schema-inspector"));

var _profileTokens = require("./profileTokens");

var _profileProofs = require("./profileProofs");

var _profileZoneFiles = require("./profileZoneFiles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const schemaDefinition = {
  type: 'object',
  properties: {
    '@context': {
      type: 'string',
      optional: true
    },
    '@type': {
      type: 'string'
    }
  }
};

class Profile {
  constructor(profile = {}) {
    this._profile = Object.assign({}, {
      '@context': 'http://schema.org/'
    }, profile);
  }

  toJSON() {
    return Object.assign({}, this._profile);
  }

  toToken(privateKey) {
    return (0, _profileTokens.signProfileToken)(this.toJSON(), privateKey);
  }

  static validateSchema(profile, strict = false) {
    schemaDefinition.strict = strict;
    return _schemaInspector.default.validate(schemaDefinition, profile);
  }

  static fromToken(token, publicKeyOrAddress = null) {
    const profile = (0, _profileTokens.extractProfile)(token, publicKeyOrAddress);
    return new Profile(profile);
  }

  static makeZoneFile(domainName, tokenFileURL) {
    return (0, _profileZoneFiles.makeProfileZoneFile)(domainName, tokenFileURL);
  }

  static validateProofs(domainName) {
    return (0, _profileProofs.validateProofs)(this.toJSON(), domainName);
  }

}

exports.Profile = Profile;
