// blockstack/lib/profiles/services/linkedIn.js
"use strict";

exports.__esModule = true;
exports.LinkedIn = void 0;

var _cheerio = _interopRequireDefault(require("cheerio"));

var _service = require("./service");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class LinkedIn extends _service.Service {
  static getBaseUrls() {
    const baseUrls = ['https://www.linkedin.com/feed/update/', 'http://www.linkedin.com/feed/update/', 'www.linkedin.com/feed/update/'];
    return baseUrls;
  }

  static getProofUrl(proof) {
    const baseUrls = this.getBaseUrls();
    let proofUrl = proof.proof_url.toLowerCase();
    proofUrl = super.prefixScheme(proofUrl);

    for (let i = 0; i < baseUrls.length; i++) {
      if (proofUrl.startsWith(`${baseUrls[i]}`)) {
        return proofUrl;
      }
    }

    throw new Error(`Proof url ${proof.proof_url} is not valid for service ${proof.service}`);
  }

  static shouldValidateIdentityInBody() {
    return true;
  }

  static getProofIdentity(searchText) {
    const $ = _cheerio.default.load(searchText);

    const profileLink = $('article').find('.post-meta__profile-link');

    if (profileLink !== undefined) {
      if (profileLink.attr('href') === undefined) {
        return '';
      }

      return profileLink.attr('href').split('/').pop();
    } else {
      return '';
    }
  }

  static getProofStatement(searchText) {
    const $ = _cheerio.default.load(searchText);

    const postContent = $('article').find('.commentary');
    let statement = '';

    if (postContent !== undefined) {
      statement = postContent.text();
    }

    return statement;
  }

}

exports.LinkedIn = LinkedIn;
