// blockstack/lib/storage/hub.js
"use strict";

exports.__esModule = true;
exports.uploadToGaiaHub = uploadToGaiaHub;
exports.getFullReadUrl = getFullReadUrl;
exports.connectToGaiaHub = connectToGaiaHub;
exports.setLocalGaiaHubConnection = setLocalGaiaHubConnection;
exports.getOrSetLocalGaiaHubConnection = getOrSetLocalGaiaHubConnection;
exports.getBucketUrl = getBucketUrl;
exports.BLOCKSTACK_GAIA_HUB_LABEL = void 0;

var _bitcoinjsLib = _interopRequireDefault(require("bitcoinjs-lib"));

var _crypto = _interopRequireDefault(require("crypto"));

var _jsontokens = require("jsontokens");

var _utils = require("../utils");

var _index = require("../index");

var _authConstants = require("../auth/authConstants");

var _userSession = require("../auth/userSession");

var _logger = require("../logger");

var _errors = require("../errors");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const BLOCKSTACK_GAIA_HUB_LABEL = 'blockstack-gaia-hub-config';
exports.BLOCKSTACK_GAIA_HUB_LABEL = BLOCKSTACK_GAIA_HUB_LABEL;

function uploadToGaiaHub(filename, contents, hubConfig, contentType = 'application/octet-stream') {
  _logger.Logger.debug(`uploadToGaiaHub: uploading ${filename} to ${hubConfig.server}`);

  return fetch(`${hubConfig.server}/store/${hubConfig.address}/${filename}`, {
    method: 'POST',
    headers: {
      'Content-Type': contentType,
      Authorization: `bearer ${hubConfig.token}`
    },
    body: contents
  }).then(response => {
    if (response.ok) {
      return response.text();
    } else {
      throw new Error('Error when uploading to Gaia hub');
    }
  }).then(responseText => JSON.parse(responseText)).then(responseJSON => responseJSON.publicURL);
}

function getFullReadUrl(filename, hubConfig) {
  return `${hubConfig.url_prefix}${hubConfig.address}/${filename}`;
}

function makeLegacyAuthToken(challengeText, signerKeyHex) {
  // only sign specific legacy auth challenges.
  let parsedChallenge;

  try {
    parsedChallenge = JSON.parse(challengeText);
  } catch (err) {
    throw new Error('Failed in parsing legacy challenge text from the gaia hub.');
  }

  if (parsedChallenge[0] === 'gaiahub' && parsedChallenge[3] === 'blockstack_storage_please_sign') {
    const signer = (0, _index.hexStringToECPair)(signerKeyHex + (signerKeyHex.length === 64 ? '01' : ''));

    const digest = _bitcoinjsLib.default.crypto.sha256(challengeText);

    const signature = signer.sign(digest).toDER().toString('hex');
    const publickey = (0, _index.getPublicKeyFromPrivate)(signerKeyHex);
    const token = Buffer.from(JSON.stringify({
      publickey,
      signature
    })).toString('base64');
    return token;
  } else {
    throw new Error('Failed to connect to legacy gaia hub. If you operate this hub, please update.');
  }
}

function makeV1GaiaAuthToken(hubInfo, signerKeyHex, hubUrl, associationToken) {
  const challengeText = hubInfo.challenge_text;
  const handlesV1Auth = hubInfo.latest_auth_version && parseInt(hubInfo.latest_auth_version.slice(1), 10) >= 1;
  const iss = (0, _index.getPublicKeyFromPrivate)(signerKeyHex);

  if (!handlesV1Auth) {
    return makeLegacyAuthToken(challengeText, signerKeyHex);
  }

  const salt = _crypto.default.randomBytes(16).toString('hex');

  const payload = {
    gaiaChallenge: challengeText,
    hubUrl,
    iss,
    salt,
    associationToken
  };
  const token = new _jsontokens.TokenSigner('ES256K', signerKeyHex).sign(payload);
  return `v1:${token}`;
}

function connectToGaiaHub(gaiaHubUrl, challengeSignerHex, associationToken) {
  _logger.Logger.debug(`connectToGaiaHub: ${gaiaHubUrl}/hub_info`);

  return fetch(`${gaiaHubUrl}/hub_info`).then(response => response.json()).then(hubInfo => {
    const readURL = hubInfo.read_url_prefix;
    const token = makeV1GaiaAuthToken(hubInfo, challengeSignerHex, gaiaHubUrl, associationToken);
    const address = (0, _utils.ecPairToAddress)((0, _index.hexStringToECPair)(challengeSignerHex + (challengeSignerHex.length === 64 ? '01' : '')));
    return {
      url_prefix: readURL,
      address,
      token,
      server: gaiaHubUrl
    };
  });
}
/**
 * These two functions are app-specific connections to gaia hub,
 *   they read the user data object for information on setting up
 *   a hub connection, and store the hub config to localstorage
 * @param {UserSession} caller - the instance calling this function
 * @private
 * @returns {Promise} that resolves to the new gaia hub connection
 */


function setLocalGaiaHubConnection(caller) {
  const userData = caller.loadUserData();

  if (!userData) {
    throw new _errors.InvalidStateError('Missing userData');
  }

  if (!userData.hubUrl) {
    userData.hubUrl = _authConstants.BLOCKSTACK_DEFAULT_GAIA_HUB_URL;
  }

  return connectToGaiaHub(userData.hubUrl, userData.appPrivateKey, userData.associationToken).then(gaiaConfig => {
    userData.gaiaHubConfig = gaiaConfig;
    return gaiaConfig;
  });
}

function getOrSetLocalGaiaHubConnection(caller) {
  const userData = caller.store.getSessionData().userData;

  if (!userData) {
    throw new _errors.InvalidStateError('Missing userData');
  }

  const hubConfig = userData.gaiaHubConfig;

  if (hubConfig) {
    return Promise.resolve(hubConfig);
  }

  return setLocalGaiaHubConnection(caller);
}

function getBucketUrl(gaiaHubUrl, appPrivateKey) {
  let challengeSigner;

  try {
    challengeSigner = _bitcoinjsLib.default.ECPair.fromPrivateKey(new Buffer(appPrivateKey, 'hex'));
  } catch (e) {
    return Promise.reject(e);
  }

  return fetch(`${gaiaHubUrl}/hub_info`).then(response => response.text()).then(responseText => JSON.parse(responseText)).then(responseJSON => {
    const readURL = responseJSON.read_url_prefix;
    const address = (0, _utils.ecPairToAddress)(challengeSigner);
    const bucketUrl = `${readURL}${address}/`;
    return bucketUrl;
  });
}
