// blockstack/lib/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  makeDIDFromAddress: true,
  makeDIDFromPublicKey: true,
  getDIDType: true,
  getAddressFromDID: true,
  getEntropy: true,
  makeECPrivateKey: true,
  publicKeyToAddress: true,
  getPublicKeyFromPrivate: true,
  nextYear: true,
  nextMonth: true,
  nextHour: true,
  makeUUID4: true,
  updateQueryStringParameter: true,
  isLaterVersion: true,
  isSameOriginAbsoluteUrl: true,
  hexStringToECPair: true,
  ecPairToHexString: true,
  ecPairToAddress: true,
  transactions: true,
  safety: true,
  TransactionSigner: true,
  PubkeyHashSigner: true,
  addUTXOsToFund: true,
  estimateTXBytes: true,
  BlockstackWallet: true,
  IdentityKeyPair: true,
  network: true,
  decodeToken: true,
  config: true,
  encryptMnemonic: true,
  decryptMnemonic: true,
  UserSession: true
};
exports.UserSession = exports.decryptMnemonic = exports.encryptMnemonic = exports.config = exports.decodeToken = exports.network = exports.IdentityKeyPair = exports.BlockstackWallet = exports.estimateTXBytes = exports.addUTXOsToFund = exports.PubkeyHashSigner = exports.TransactionSigner = exports.safety = exports.transactions = exports.ecPairToAddress = exports.ecPairToHexString = exports.hexStringToECPair = exports.isSameOriginAbsoluteUrl = exports.isLaterVersion = exports.updateQueryStringParameter = exports.makeUUID4 = exports.nextHour = exports.nextMonth = exports.nextYear = exports.getPublicKeyFromPrivate = exports.publicKeyToAddress = exports.makeECPrivateKey = exports.getEntropy = exports.getAddressFromDID = exports.getDIDType = exports.makeDIDFromPublicKey = exports.makeDIDFromAddress = void 0;

var _queryString = _interopRequireDefault(require("query-string"));

var _auth = require("./auth");

Object.keys(_auth).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _auth[key];
});

var _profiles = require("./profiles");

Object.keys(_profiles).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _profiles[key];
});

var _storage = require("./storage");

Object.keys(_storage).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _storage[key];
});

var _dids = require("./dids");

exports.makeDIDFromAddress = _dids.makeDIDFromAddress;
exports.makeDIDFromPublicKey = _dids.makeDIDFromPublicKey;
exports.getDIDType = _dids.getDIDType;
exports.getAddressFromDID = _dids.getAddressFromDID;

var _keys = require("./keys");

exports.getEntropy = _keys.getEntropy;
exports.makeECPrivateKey = _keys.makeECPrivateKey;
exports.publicKeyToAddress = _keys.publicKeyToAddress;
exports.getPublicKeyFromPrivate = _keys.getPublicKeyFromPrivate;

var _utils = require("./utils");

exports.nextYear = _utils.nextYear;
exports.nextMonth = _utils.nextMonth;
exports.nextHour = _utils.nextHour;
exports.makeUUID4 = _utils.makeUUID4;
exports.updateQueryStringParameter = _utils.updateQueryStringParameter;
exports.isLaterVersion = _utils.isLaterVersion;
exports.isSameOriginAbsoluteUrl = _utils.isSameOriginAbsoluteUrl;
exports.hexStringToECPair = _utils.hexStringToECPair;
exports.ecPairToHexString = _utils.ecPairToHexString;
exports.ecPairToAddress = _utils.ecPairToAddress;

var _operations = require("./operations");

exports.transactions = _operations.transactions;
exports.safety = _operations.safety;
exports.TransactionSigner = _operations.TransactionSigner;
exports.PubkeyHashSigner = _operations.PubkeyHashSigner;
exports.addUTXOsToFund = _operations.addUTXOsToFund;
exports.estimateTXBytes = _operations.estimateTXBytes;

var _wallet = require("./wallet");

exports.BlockstackWallet = _wallet.BlockstackWallet;
exports.IdentityKeyPair = _wallet.IdentityKeyPair;

var _network = require("./network");

exports.network = _network.network;

var _jsontokens = require("jsontokens");

exports.decodeToken = _jsontokens.decodeToken;

var _config = require("./config");

exports.config = _config.config;

var _encryption = require("./encryption");

exports.encryptMnemonic = _encryption.encryptMnemonic;
exports.decryptMnemonic = _encryption.decryptMnemonic;

var _userSession = require("./auth/userSession");

exports.UserSession = _userSession.UserSession;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO: Putting in here so it executes ASAP. There is probably a better place to put this.
// Note: This prototype is designed to work as a drop-in-replacement (non-breaking upgrade)
// for apps using blockstack.js. That requires doing this hacky global & immediate detection. 
// A more proper approach would require developers to call an additional blockstack.js method 
// for invoking this detection method.
(function protocolEchoReplyDetection() {
  // Check that the `window` APIs exist
  if (typeof window !== 'object' || !window.location || !window.localStorage) {
    // Exit detection function - we are not running in a browser environment.
    return;
  } // Check if the location query string contains a protocol-echo reply.
  // If so, this page was only re-opened to signal back the originating 
  // tab that the protocol handler is installed. 


  const queryDict = _queryString.default.parse(window.location.search);

  if (queryDict.echoReply) {
    // Use localStorage to notify originated tab that protocol handler is available and working.
    const echoReplyKey = `echo-reply-${queryDict.echoReply}`; // Set the echo-reply result in localStorage for the other window to see.

    window.localStorage.setItem(echoReplyKey, 'success'); // Redirect back to the localhost auth url, as opposed to another protocol launch.
    // This will re-use the same tab rather than creating another useless one.

    window.setTimeout(() => {
      window.location = decodeURIComponent(queryDict.authContinuation);
    }, 10);
  }
})();
